storage.go

```diff
@@ -0,0 +1,342 @@
+package dynamodbstorage
+
+import (
+	"encoding/base64"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
+
+	"github.com/caddyserver/certmagic"
+)
+
+const (
+	contentsAttribute    = "Contents"
+	primaryKeyAttribute  = "PrimaryKey"
+	lastUpdatedAttribute = "LastUpdated"
+	lockTimeoutMinutes   = 5 * time.Minute
+	lockPollingInterval  = 5 * time.Second
+)
+
+// Item holds structure of domain, certificate data,
+// and last updated for marshaling with DynamoDb
+type Item struct {
+	PrimaryKey  string    `json:"PrimaryKey"`
+	Contents    string    `json:"Contents"`
+	LastUpdated time.Time `json:"LastUpdated"`
+}
+
+// Storage implements certmagic.Storage to facilitate
+// storage of certificates in DynamoDB for a clustered environment.
+// Also implements certmagic.Locker to facilitate locking
+// and unlocking of cert data during storage
+type Storage struct {
+	Table               string
+	AwsSession          *session.Session
+	AwsEndpoint         string
+	AwsRegion           string
+	AwsDisableSSL       bool
+	LockTimeout         time.Duration
+	LockPollingInterval time.Duration
+}
+
+// initConfig initializes configuration for table nam and AWS session
+func (s *Storage) initConfig() error {
+	if s.Table == "" {
+		return fmt.Errorf("config error: table name is required")
+	}
+
+	if s.LockTimeout == 0*time.Second {
+		s.LockTimeout = lockTimeoutMinutes
+	}
+	if s.LockPollingInterval == 0*time.Second {
+		s.LockPollingInterval = lockPollingInterval
+	}
+
+	// Initialize AWS Session if needed
+	if s.AwsSession == nil {
+		var err error
+		s.AwsSession, err = session.NewSession(&aws.Config{
+			Endpoint:   &s.AwsEndpoint,
+			Region:     &s.AwsRegion,
+			DisableSSL: &s.AwsDisableSSL,
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Store puts value at key.
+func (s *Storage) Store(key string, value []byte) error {
+	if err := s.initConfig(); err != nil {
+		return err
+	}
+
+	encVal := base64.StdEncoding.EncodeToString(value)
+
+	if key == "" {
+		return fmt.Errorf("key must not be empty")
```

Comment:
If accepted, remember to import `errors`. Other similar instances can also be changed if desired.
```suggestion
		return errors.New("key must not be empty")
```

Reply:
sure, thanks